create wallet standard aip (#305)
aptos-foundation/AIPs: aips/wallet-standard.md

---
aip: (this is determined by the AIP Manager, leave it empty when drafting)
title: Wallet Standard
author: 0xmaayan, hardsetting, NorbertBodziony
discussions-to (*optional): https://github.com/aptos-foundation/AIPs/issues/306
Status: Draft
last-call-end-date (*optional): 09/02/2024
type: Standard
created: 29/01/2024
---

# AIP-X - Wallet Standard

## Summary

> A Wallet Standard is an API interface a wallet implements and defines how applications interact with a wallet. This AIP introduces a new wallet standard for the Aptos ecosystem.

Reviewer: ```suggestion
The Wallet standard defines a universal API for wallet and application interactions. This AIP introduces a new wallet standard for the Aptos ecosystem.
```

- Register the wallet so it will be discoverable by the dapp
- Implementation of a AptosWallet class that conforms with the new standard

## Rationale

This [chain agnostic solution](https://github.com/wallet-standard/wallet-standard) has been introduced as a generalized standard for wallets and dapps communication and has already been implemented on [Solana](https://github.com/wallet-standard/wallet-standard) and [Sui](https://docs.sui.io/standards/wallet-standard), and the [Ethereum](https://eips.ethereum.org/EIPS/eip-6963) community recently proposed a similar solution.

## Specification

The [Wallet Standard](https://github.com/aptos-labs/wallet-standard) is a chain-agnostic set of interfaces and conventions that aim to improve how applications interact with injected wallets.

**Standard Features**

A standard feature is a method that must or should be supported and implemented by a wallet.
Here is a list of the suggested [Aptos features](https://github.com/aptos-labs/wallet-standard/tree/main/src/features)

> a feature marked with `*` is an optional feature

`aptos:connect` method to establish a connection between a dapp and a wallet.

```ts

> // `silent?: boolean` - gives ability to trigger connection without user prompt (for example, for auto-connect)

Author: I believe this is for autoConnect functionality. maybe the user has given the wallet permission to auto connect and doesnt want to aprrove the popup everytime. Same feature as one of the wallets have - i think it is Fewcha?

> // `silent?: boolean` - gives ability to trigger connection without user prompt (for example, for auto-connect)

Reviewer 2: oh I see, wasn't aware we wanted to allow that

// `networkInfo?: NetworkInfo` - defines the network that the dapp will use (shortcut for connect and change network)

connect(silent?: boolean, networkInfo?: NetworkInfo): Promise<UserResponse<AccountInfo>>;
```

`aptos:disconnect` method to disconnect a connection established between a dapp and a wallet

```ts
disconnect(): Promise<void>;
```

`aptos:getAccount` to get the current connected account in the wallet

```ts
getAccount():Promise<UserResponse<AccountInfo>>
```

`aptos:getNetwork` to get the current network in the wallet

```ts
getNetwork(): Promise<UserResponse<NetworkInfo>>;
```

> `aptos:signAndSubmitTransaction` method to sign and submit a transaction using the current connected account in the wallet.

Reviewer 2: I think this should be optional.
The same functionality can be achieved by an `aptos:signTransaction` + submitting the transaction using the AptosClient.

The new Aptos Wallet Adapter can easily provide a fallback implementation.

Author: I also think a wallet should only supports `sign` functionality. no `signAndSubmitTransaction` or `submitTransaction`. will make this optional for a wallet to implement

```

`aptos:openInMobileApp*` a function that supports redirecting a user from a web browser on mobile to a native mobile app. The wallet plugin should add the location url a wallet should open the in-app browser at.

```ts
openInMobileApp(): void
```

Types

> Note: `UserResponse` type is used for when a user rejects a rejectable request. For example, when user wants to connect but instead closes the window popup.

```ts
export interface UserApproval<TResponseArgs> {
 status: 'approved'
 args: TResponseArgs
}

export interface UserRejection {
 status: 'rejected'
}

export type UserResponse<TResponseArgs> = UserApproval<TResponseArgs> | UserRejection;

export interface AccountInfo = { account: Account, ansName?: string }

export interface NetworkInfo {
  name: Network
  chainId: number
  url?: string
}

export type AptosSignMessageInput = {
  address?: boolean
  application?: boolean
  chainId?: boolean
  message: string
  nonce: string
}

export type AptosSignMessageOutput = {
  address?: string
  application?: string
  chainId?: number
  fullMessage: string
  message: string
  nonce: string
  prefix: 'APTOS'
  signature: Signature | Signature[]

>   bitmap?: Uint8Array

Reviewer 2: I think this should be just `Signature`.
Both `MultiEd25519Signature` and `MultiKeySignature` already contain a bitmap field.

Author: @hardsetting do you mean we should remove the `bitmap` prop or change its type to `Signature`?

Reviewer 2: I'm guessing `Signature` here is the abstract class / union type for _any_ signature
i.e. `Ed25519Signature | MultiEd25519Signature | AnySignature | MultiKeySignature`

since `MultiEd25519Signature` and `MultiKeySignature` already contain a field for `bitmap` and `signatures`, I would change this like such:

```suggestion
  signature: Signature
```

> }

Reviewer 2: I think one of the first priorities after getting this standard out would be to work on a message signature standard, that includes domain separation and allow structured messages (we've been getting lots of requests about this).
Alin should probably be pulled for this

```

<ins>Register Events</ins>

On first load, and before the dapp has been loaded, it gets all the wallets that have been registered so far. To keep getting all the registered wallets after this point, the dapp must add an event listener for new wallets that get registered receiving an unsubscribe function, which it can later use to remove the listener.

```ts
const removeRegisterListener = on("register", function () {
  // The dapp can add new aptos wallets to its own state context as they are registered
  let { aptosWallets } = getAptosWallets();
});

const removeUnregisterListener = on("unregister", function () {
  let { aptosWallets } = getAptosWallets();
});
```

The dapp has an event listener now, so it sees new wallets immediately and doesn't need to poll or list them again.
This also works if the dapp loads before any wallets (it will initialize, see no wallets, then see wallets as they load)

<ins>Wallet Request</ins>

A dapp makes a wallet request by calling the feature name that coresponds to the desired action.

```ts
const onConnect = () => {
  this.wallet.features["aptos:connect"].connect();
};
```

## Risks and Drawbacks

> The new standard uses the [new TypeScript SDK](https://github.com/aptos-labs/aptos-ts-sdk) types and therefore requires dapps and wallets to use/migrate to the new TypeScript SDK or hold a conversion layer from the new TypeScript SDK types to the legacy TypeScript SDK types.

Reviewer 2: Just throwing this out there, but it'd be really cool to have a migration tool, that automatically converts your code to use the new SDK.

I know it's not trivial though.

Author: agreed. the adapter (and I believe also Petra plugin) already uses this kind of a conversion layer, we can get it into its own package to be used by wallets
